### DISCUSSION TITLE
Does `Microsoft.Extensions.DependencyInjection` support nested scopes?

### DISCUSSION THREAD

**Discussion by MartyIX** [dotnet/runtime#104887] [Q&A] [answered]
I have read about scopes in Microsoft DI and it seems that the main use case is for ASP.NET where a scope corresponds with handling of a single HTTP request.

However, does Microsoft DI support adding a child scope to a scope?[^1][^2]

[^1]: I have researched it more but the simple question makes it easier to answer.
[^2]: AFAIK, Autofac can do it but I would like to avoid 3rd party library use if possible.

**Comment by andriivitiv** [comment]
It does. You could use the [`CreateScope`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.serviceproviderserviceextensions.createscope?view=net-8.0) extension method.

```cs
using var provider = new ServiceCollection()
    .AddScoped<Foo>()
    .BuildServiceProvider();
            
using (var scope = provider.CreateScope())
{
    var foo = scope.ServiceProvider.GetRequiredService<Foo>();
}
```

However, Autofac is more configurable, e.g. you could register additional services when creating a child scope.

**Reply by MartyIX** [reply]
I see but how about making it really *nested*:

```cs
using var provider = new ServiceCollection()
    .AddScoped<Foo>()
    .BuildServiceProvider();
            
using (var scope = provider.CreateScope())
{
    var foo = scope.ServiceProvider.GetRequiredService<Foo>();

    using (var scope2 = scope.ServiceProvider.CreateScope())
    {
        var foo2 = scope2.ServiceProvider.GetRequiredService<Foo>(); 
    }
}
```

Is this supported and if so how does it behave? I guess I should create a small sample and try it out. However, having it documented would be great because then I could verify it against official doc.

**Reply by wvpm** [reply]
As your example points out, you can create scopes from both the (root) `provider` and the `scope.ServiceProvider`.
There is no difference between using the root provider to create a scope or using the scoped provider for that.
You simply create a new scope. There is no stacktrace for scopes or any sort of nesting.
It's either you're in the root scope and should not resolve scoped instances (validating scopes will fails).
Or you're in a scope and you can resolve anything (Singleton/Scoped/Transient).

There are still use cases for your code example. Imagine `scope` is an HTTP request level scope and you query some data.
You can then split the data in chunks and use `scope2` to resolve services that process each chunk.

**Reply by MartyIX** [reply]
> There are still use cases for your code example. Imagine `scope` is an HTTP request level scope and you query some data.
> You can then split the data in chunks and use `scope2` to resolve services that process each chunk.

Could you clarify this part? I'm a bit confused what one would achieve with that precisely.

**Reply by wvpm** [reply]
I vaguely remember code that was used in bulk operations. A scope would be created per trigger (timer I believe). That scope would retrieve the data to be processed. The data would be split into equal sized chunks and a new scope would be created for each chunk. The chunks of data would then be processed using services from that scope.

I no longer have access to the code and my memory of it is vague. It still demonstrates a scenario where you might create scopes explicitly.

As for nesting, the trigger scope and the data chunk scope siblings, not parent-child related.

**Comment by davidfowl** [comment, marked as answer]
There's no support for nested scopes.
